src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -0,0 +1,146 @@
+package is.hail.expr.ir
+
+import is.hail.asm4s._
+import is.hail.asm4s.TypeInfo
+import is.hail.expr.Type
+import is.hail.expr.TStruct
+import is.hail.annotations._
+import org.objectweb.asm.tree._
+import org.objectweb.asm.Opcodes._
+
+import scala.language.existentials
+
+object Compile {
+  case class DetailedTypeInfo[T, UT : TypeInfo](injectNonMissing: Option[(TypeInfo[T], Code[_] => Code[_])]) {
+    val uti = typeInfo[UT]
+  }
+
+  class MissingBits(fb: FunctionBuilder[_]) {
+    private var used = 0
+    private var bits: LocalRef[Int] = null
+
+    def newBit(x: Code[Boolean]): Code[Boolean] = {
+      if (used >= 64 || bits == null) {
+        bits = fb.newLocal[Int]
+        fb.emit(bits.store(0))
+        used = 0
+      }
+
+      fb.emit(bits.store(bits | (x.asInstanceOf[Code[Int]] << used)))
+
+      val read = bits & (1 << used)
+      used += 1
+      read.asInstanceOf[Code[Boolean]]
+    }
+  }
+
+  def apply(x: IR, outTyps: Array[DetailedTypeInfo[_,_]], fb: FunctionBuilder[_]) {
+    terminal(x, Map(), outTyps, fb, new MissingBits(fb))
+  }
+
+  private[ir] def nonTerminal(x: IR, env: Map[String, (Code[Boolean], Code[_])], outTyps: Array[DetailedTypeInfo[_,_]], fb: FunctionBuilder[_], mb: MissingBits): (Code[Boolean], Code[_]) = {
```

Comment:
I think terminal and non-terminal are non-standard.  What about expression and statement?